File: FunnyRoad/app/src/main/java/com/bestdeveloper/funnyroad/activity/WelcomeActivity.java
package com.bestdeveloper.funnyroad.activity;

import static android.content.ContentValues.TAG;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.bestdeveloper.funnyroad.R;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.FirebaseFirestore;

public class WelcomeActivity extends AppCompatActivity {

    FirebaseAuth mAuth;

    // Buttons

    Button jSignIn;
    Button jSignUp;

    EditText loginEmail;
    EditText loginPassword;

    // Connection to Firebase
    private FirebaseFirestore db = FirebaseFirestore.getInstance();
    private CollectionReference collectionReference = db.collection("Users");
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_welcome);

        mAuth = FirebaseAuth.getInstance();

        jSignIn = findViewById(R.id.welcome_signInBtn);
        jSignUp = findViewById(R.id.welcome_signUpBtn);
        loginEmail = findViewById(R.id.loginEmail);
        loginPassword = findViewById(R.id.loginPassword);

        jSignIn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                signIn(loginEmail.getText().toString().trim(), loginPassword.getText().toString().trim());
            }
        });

        jSignUp.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent i = new Intent(WelcomeActivity.this, SignUpActivity.class);
                startActivity(i);
            }
        });


    }

    private void signIn(String email, String password){
        if(!TextUtils.isEmpty(email) && !TextUtils.isEmpty(password)) {
            mAuth.signInWithEmailAndPassword(email, password)
                    .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                        @Override
                        public void onComplete(@NonNull Task<AuthResult> task) {
                            if (task.isSuccessful()) {
                                // Sign in success, update UI with the signed-in user's information
                                Log.d(TAG, "signInWithEmail:success");
                                FirebaseUser user = mAuth.getCurrentUser();
                                startActivity(new Intent(WelcomeActivity.this, MapActivity.class));
                            } else {
                                // If sign in fails, display a message to the user.
                                Log.w(TAG, "signInWithEmail:failure", task.getException());
                                Toast.makeText(WelcomeActivity.this, "Authentication failed.",
                                        Toast.LENGTH_SHORT).show();
                            }
                        }
                    });
        }

    }

    @Override
    protected void onStart() {
        super.onStart();
        if(mAuth.getCurrentUser() != null){
            Log.i("TAG", "user: " + mAuth.getCurrentUser().getUid());
            startActivity(new Intent(this, MapActivity.class));
        }

    }
}

File: FunnyRoad/app/src/main/java/com/bestdeveloper/funnyroad/activity/WelcomeActivity.kt
package com.bestdeveloper.funnyroad.activity

import EditTextUtils
import EditTextUtils.highlightFieldAndSetError
import EditTextUtils.setDefaultStrokeOnChangedEditText
import android.content.ContentValues
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.bestdeveloper.funnyroad.R
import com.bestdeveloper.funnyroad.databinding.ActivityWelcomeBinding

class WelcomeActivity : AppCompatActivity() {
    private lateinit var load_txt: TextView
    private lateinit var mAuth: FirebaseAuthenticator

    // Binding
    private lateinit var binding: ActivityWelcomeBinding

    private lateinit var progressBar: ProgressBar
    private lateinit var greyBackground: View

    private val navigator = ViewNavigator()

    // view model
    private lateinit var viewModel : WelcomeActivityViewModel



    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityWelcomeBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = ViewModelProvider(this).get(WelcomeActivityViewModel::class.java)

        mAuth = FirebaseAuthenticator()
        progressBar = binding.logInPrBar
        greyBackground = binding.lowTrBackground
        load_txt = binding.loggingInTextBar
        switchBars(View.GONE)
        
        setDefaultStrokeOnChangedEditText(applicationContext, binding.loginEmail)
        setDefaultStrokeOnChangedEditText(applicationContext, binding.loginPassword)

        // Set up click listeners using binding
        binding.logInBtn.setOnClickListener {
            val email = EditTextUtils.trimInput(binding.loginEmail)
            val password = EditTextUtils.trimInput(binding.loginPassword)
            validate(email, password, binding.incPassString)
        }

        binding.signUpTxtBtn.setOnClickListener {
            navigator.navigateToSignUpActivity(this)
        }

    }

    private fun validate(email: String, password: String, errorText: TextView) {
        CredentialValidator().validate(email, password, object : ValidationInterface{
            override fun emailOnEmptyResult() {
                highlightFieldAndSetError(applicationContext, binding.loginEmail, errorText, getString(R.string.please_enter_your_email))
            }

            override fun passwordOnEmptyResult() {
                highlightFieldAndSetError(applicationContext, binding.loginPassword, errorText, getString(R.string.please_enter_your_password))
            }

            override fun onBothEmpty() {
                highlightFieldAndSetError(applicationContext, binding.loginEmail, errorText, "")
                highlightFieldAndSetError(applicationContext, binding.loginPassword, errorText, "")
            }

            override fun emailOnInvalidResult() {
                highlightFieldAndSetError(applicationContext, binding.loginEmail, errorText,  getString(R.string.incorrect_email))
            }

            override fun passwordOnInvalidResult() {
               highlightFieldAndSetError(applicationContext, binding.loginPassword, errorText, getString(R.string.short_password))
            }



            override fun OnSuccessResult() {
                switchBars(View.VISIBLE)
                viewModel.signIn(email, password, object : OnSignUpResult{
                    override fun onSuccess() {
                        onLogInSuccess()
                    }

                    override fun onFailure() {
                        onLogInFailure()
                    }
                })
            }
        })

    }


    private fun onLogInSuccess(){
        switchBars(View.GONE)
        binding.incPassString.text = ""
        Log.d(ContentValues.TAG, "signInWithEmail:success")
        navigator.navigateToMapActivity(this)
    }

    private fun onLogInFailure(){
        switchBars(View.GONE)
        binding.incPassString.text = resources.getString(R.string.incorrect_credentials)
        Log.w(ContentValues.TAG, "signInWithEmail:failure")
        Toast.makeText(
            this@WelcomeActivity, "Authentication failed.",
            Toast.LENGTH_SHORT
        ).show()
    }


    override fun onStart() {
        super.onStart()
        if (viewModel.getCurrentUser() == null) {
            Log.i("TAG", "user: " + viewModel.getCurrentUser()!!.uid)
            navigator.navigateToMapActivity(this)
        }
    }

    private fun switchBars(viewState : Int) {
        load_txt.visibility = viewState
        progressBar.visibility = viewState
        greyBackground.visibility = viewState
    }
}
